src/liballoc/collections/btree/node.rs

```diff
@@ -531,16 +531,11 @@ impl<'a, K: 'a, V: 'a, Type> NodeRef<marker::Immut<'a>, K, V, Type> {
             // because we might be the shared root.
             // For this reason, `NodeHeader` has this `K2` parameter (that's usually `()`
             // and hence just adds a size-0-align-1 field, not affecting layout).
-            // We know that we can transmute `NodeHeader<K, V, ()>` to `NodeHeader<K, V, K>`
-            // because we did the alignment check above, and hence `NodeHeader<K, V, K>`
-            // is not bigger than `NodeHeader<K, V, ()>`!  Then we can use `NodeHeader<K, V, K>`
+            // We can transmute `NodeHeader<K, V, ()>` to `NodeHeader<K, V, K>` and use that
```

Comment:
Well but why can we do that without leaving the bounds of the allocation? If this is the shared root, it's not big enough for `NodeHeader<K, V, K>`, right? That's exactly what the assertion was testing, and I don't see how just removing it could be correct.

Reply:
Why can we transmute? I'm not even sure what transmuting is (I'm a recovering C++ addict), I just tried to preserve as much of the comment while remaining truthful. But I think here it's because we really cast to LeafNode which is even better than NodeHeader<K, V, K>.

If this is the shared root, then it is big enough for NodeHeader<K, V, K> because we're in the else-part of `mem::align_of::<K>() > mem::align_of::<LeafNode<(), ()>>()  && self.is_shared_root() `. I guess that test on K's alignment comes down to the same thing as checking sizes? Or maybe you meant to write the assertion in the then-part instead of the else-part?

Reply:
No, wait, I meant this: we need to insert `|| !self.is_shared_root()` into the original assert to be correct and to prove that K's alignment check already makes sure it's big enough. But obviously that would not be "completely compile-time evaluated".

Reply:
> I'm not even sure what transmuting is (I'm a recovering C++ addict)

It's basically the same as a `reinterpret_cast`.

You are taking some data that has type `NodeHeader<K, V, ()>` and treat it as if it has `NodeHeader<K, V, K>`. In particular, this creates a *reference* of the latter type. A reference must not be dangling, meaning the memory range it points to (as determined by the size of the pointee) must be in-bounds of an allocation. If this is the static root, `NodeHeader<K, V, ()>`  might be in-bounds but `NodeHeader<K, V, K>`, if bigger, might not be in-bounds any more. That's exactly what the alignment check is supposed to protect against.

Reply:
> Isn't it better then to write exactly what we want in the test? As in:

That'll still fail the assert though, right?

I agree with your earlier statement, the assertion should be
```rust
assert!(!self.is_shared_root() || mem::size_of::<NodeHeader<K, V>>() == mem::size_of::<NodeHeader<K, V, K>>());
```
IOW, *if* this is the shared root, then the size stuff works out. If it's not the shared root we do not need to be afraid of anything as we actually have a properly sized allocation.

This would have to become a `debug_assert!` though, as it now would be a run-time thing.

Reply:
> > Isn't it better then to write exactly what we want in the test? As in:
> 
> That'll still fail the assert though, right?

Well, it would fail the old flawed assert, but not the corrected assert which is the exact opposite of that condition, so it would be pointless to repeat it as an assert.

Reply:
> Well, it would fail the old flawed assert, but not the corrected assert which is the exact opposite of that condition, so it would be pointless to repeat it as an assert.

Oh I see. Makes sense.

Reply:
> even if their size is the same, two types can still have different alignment, right? But not these two repr(C) types with the same initial field, I guess.

No, that's not the right interpretation. When size is different due to the `keys_start` field's alignment, the whole struct's alignment is different too. Next guess: the fact that `keys_start` is a zero size array, implies that only K's alignment influences size, not K's size.

But I realized I don't have to wear my brain guessing: I just wrote both conditions in the if-clause, the compiler spends a nanosecond comparing, and anyone reading the code can be convinced in seconds that `keys_start` in the else-part will be valid. How about that?

Reply:
Which also meant I could simply revert this part of the comments.

Reply:
Aaargh... which also meant I should copy the condition to `into_key_slice_mut`.